tools/redcap2linst.php

```diff
@@ -0,0 +1,166 @@
+    if (!empty($linstFormat)) {
+        $instruments[$inst][] = $linstFormat;
+    }
+}
+fclose($fp);
+
+outputFiles($outputDir, $instruments);
+
+/**
+ * Take a single line from the redcap dictionary and returns the
+ * closest LINST equivalent.
+ *
+ * @param string $redcaptype      The type from REDCap
+ * @param string $redcapfieldname The fieldname from REDCap
+ * @param string $redcaplabel     The label from REDCap
+ * @param string $redcapChoices   The choices columnfrom REDCap
+ *
+ * @return string
+ */
+function toLINST(
+    string $redcaptype,
+    string $redcapfieldname,
+    string $redcaplabel,
+    string $redcapChoices,
+) : string {
+    $label = str_replace($redcaplabel, "\n", "<br /><br />");
+    switch ($redcaptype) {
+    case 'text':
+        // text maps directly to LORIS
+        return "text{@}$redcapfieldname{@}$label";
+    case 'descriptive':
+        // descriptive maps to label with no field.
+        return "static{@}{@}$label";
+    case 'radio':
+    case 'dropdown':
+        // Radio or dropdown maps to a select and the options are in the
+        // same format in the dictionary.
+        return "select{@}$redcapfieldname{@}$label{@}"
+            . optionsToLINST($redcapChoices);
+    case 'checkbox':
+        // checkboxes are the same format as radios but allow multiple options,
+        // so map to a selectmultiple instead of a select
+        return "selectmultiple{@}$redcapfieldname{@}$label{@}"
+            . optionsToLINST($redcapChoices);
+    case 'yesno':
+        // Map yes/no fields to dropdowns with yes and no options.
+        return "select{@}$redcapfieldname{@}$label{@}"
+            . "NULL=>''{-}'yes'=>'Yes'{-}'no'=>'No'";
+    case 'calc':
+        // Calc maps to a score field. We create the DB field but don't do the score.
+        return "static{@}$redcapfieldname{@}$label";
+    case 'sql':
+        // The "SQL" data type seems to just be for presentation? I hope?
+        return "";
+    case 'notes':
+        // REDCap calls textareas notes
+        return "textarea{@}$redcapfieldname{@}$label";
+    default:
+        throw new \LorisException("Unhandled REDCap type $redcaptype");
+    }
+}
+
+/**
+ * Take the options column from a dictionary line and convert it to the
+ * linst format for select/multiselect
+ *
+ * @param string $dictionary The dictionary from REDCap
+ *
+ * @return string
+ */
+function optionsToLINST(string $dictionary) : string
+{
+    $choices      = explode(' | ', $dictionary);
+    $linstChoices = [];
+    foreach ($choices as $choice) {
+        $matches = [];
+        if (preg_match("/^(\s)*(\d+)(\s)*,(.*)$/", $choice, $matches) !== 1) {
+            throw new \DomainException("Could not parse radio option: $choice");
+
+        }
+        $backend        = $matches[2] . '_'
+                . preg_replace("/\s+/", "_", trim($matches[4]));
+        $linstFormat    = "'$backend'=>'" . trim(strtolower($matches[4])) . '\'';
```

Comment:
the way you've done is a sensible choice i think, e.g.:

```
select{@}method_questionnaire16{@}<br /><br />{@}'1_Online/En_ligne'=>'online/en ligne'{-}'2_Over_the_phone/Par_téléphone'=>'over the phone/par téléphone'{-}'3_In-person_during_a_study_visit/En_personne_lors_d'une_visite_d'étude'=>'in-person during a study visit/en personne lors d'une visite d'étude'{-}'4_Over_videoconference/Par_vidéconférence'=>'over videoconference/par vidéconférence'
```

however in COPN's case, they really wanted the option mappings to be this way i.e.:

```
select{@}method_questionnaire16{@}<br /><br />{@}'1'=>'online/en ligne'{-}'2'=>'over the phone/par téléphone'{-}'3'=>'in-person during a study visit/en personne lors d'une visite d'étude'{-}'4'=>'over videoconference/par vidéconférence'
```
with the option labels not made to all be lower-case

Reply:
There's some decisions to make here that maybe we can discuss at the next LORIS meeting. My understanding is that in REDCap all values are integers (which makes it easier for them to do their scoring, I guess), so we need to decide if we want to keep that, or whether we want to map it to something more descriptive for querying.